rbac: Add missing cluster-reader verbs

oshoval opened the pull request:
**What this PR does / why we need it**:
When using Openshift, in case cluster-reader
ClusterRole exists,
Users of cluster-readers group should be able to
access k8s nmstate crds, such as nns, nncp and nnce.

Add a ClusterRole that allows it in case
cluster-reader exists.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2022745

**Special notes for your reviewer**:

**Release note**:
```release-note
None
```

oshoval commented:
pretty messy, don't think it is due to this PR
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_kubernetes-nmstate/1052/pull-kubernetes-nmstate-e2e-handler-k8s/1514604352960991232

/test pull-kubernetes-nmstate-e2e-handler-k8s

qinqon commented:
/hold
This should fixed at kubernetes-nmstate project CNAO main branch no longer install it.

qinqon commented on controllers/operator/nmstate_controller.go:
```diff
@@ -153,6 +155,12 @@ func (r *NMStateReconciler) applyRBAC(instance *nmstatev1.NMState) error {
 	data.Data["HandlerPullPolicy"] = os.Getenv("HANDLER_IMAGE_PULL_POLICY")
 	data.Data["HandlerPrefix"] = os.Getenv("HANDLER_PREFIX")
+
+	err := setClusterReaderExist(r.Client, data)
+	if err != nil {
+		return errors.Wrap(err, "failed checking if cluster-reader ClusterRole exists")
+	}
```
```suggestion
	if err := setClusterReaderExist(r.Client, data); err != nil {
		return errors.Wrap(err, "failed checking if cluster-reader ClusterRole exists")
	}
```

qinqon commented on deploy/handler/cluster_role.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: k8s-nmstate-project
```
The ` is` implicit and `project`  means nothing.
```suggestion
  name: nmstate-cluster-reader
```

qinqon reviewed (changes requested):
Can we add a e2e tests u/s simulating openshit behaviour ? would be good test to run also by openshift operator owners.

oshoval commented:
> Can we add a e2e tests u/s simulating openshit behaviour ? would be good test to run also by openshift operator owners.

Removed the file from U/S it was only for checking it locally,
in order to check the feature we need openshift where we have the cluster-reader role
There we will need to do the following

1. Create a service account
2. bind the SA to the cluster-reader group
3. login as this SA
4. get nns

The PR creates the new ClusterRole, and since it has the special label 
`rbac.authorization.k8s.io/aggregate-to-cluster-reader`
Openshift will create a ClusterRoleBinding for us that binds between the new ClusterRole and the cluster-reader

beside that - addressed comments, thanks

oshoval commented:
/test pull-kubernetes-nmstate-e2e-operator-k8s

oshoval commented:
rebased

qinqon commented:
/hold cancel

oshoval commented:
Thanks Quique, will ask Openshift Operator people to review it

creydr commented on deploy/handler/cluster_role.yaml:
```diff
@@ -0,0 +1,19 @@
+- apiGroups:
+  - nmstate.io
+  resources:
+  - nodenetworkstates
+  - nodenetworkconfigurationpolicies
+  - nodenetworkconfigurationenactments
```
Adding `nmstates` (`nmstates.nmstate.io`) to this role could make sense too so we have all the nmstate.io resources, or WDYT?

oshoval replied:
It seems that we can already read it anyhow by the cluster-reader
so do we still want to add it ? (it would be duplicate, unless we remove it from somewhere else that allows it)

creydr replied:
hmm. ok. ack

creydr commented on controllers/operator/nmstate_controller.go:
```diff
@@ -153,6 +155,11 @@ func (r *NMStateReconciler) applyRBAC(instance *nmstatev1.NMState) error {
 	data.Data["HandlerPullPolicy"] = os.Getenv("HANDLER_IMAGE_PULL_POLICY")
 	data.Data["HandlerPrefix"] = os.Getenv("HANDLER_PREFIX")
+
+	if err := setClusterReaderExist(r.Client, data); err != nil {
```
IMHO it would be cleaner to have this function only return if the cluster-reader-role exists (and an error eventually) and do the `data.Data["ClusterReaderExists"] = clusterReaderRoleExists` here, so we have it aligned with the other template settings. But that might be a personal opinion...

oshoval replied:
Up to you, I started by that, but then it needs to check the error and do the assign, it looked less good imho
this way it is more compact, even that it is a bit less symmetric

oshoval commented:
The failures doesn't seem related

oshoval commented:
Added test that we can read NNS with SA that is bound to cluster-reader

qinqon commented on cluster/cluster_reader.yaml:
```diff
@@ -0,0 +1,28 @@
+---
```
This has to be under "test" not under cluster.

oshoval replied:
due to the reasons mentioned at https://github.com/nmstate/kubernetes-nmstate/pull/1052#discussion_r859554382
i think it is more suitable as part of cluster, as it prefer the cluster (independent of cluster-sync and even should be before that)

qinqon commented on cluster/up.sh:
```diff
@@ -28,3 +28,5 @@ for node in $(./cluster/kubectl.sh get nodes --no-headers | awk '{print $1}'); d
     ./cluster/cli.sh ssh ${node} -- sudo systemctl restart NetworkManager
 done
+
+./cluster/kubectl.sh create -f ./cluster/cluster_reader.yaml
```
This is part of the BearEach at tests

oshoval replied:
The cluster-reader creation part of it must be part of the cluster-up before we cluster-sync,
else the reconcile won't create the cluster-role

beside that having a test-user doesn't depend on the `cluster-sync` and might be helpful for other uses as well
as it simulates OCP env that have cluster-reader
and doesn't need to recreated again and again imo

qinqon commented on test/e2e/handler/nodes_test.go:
```diff
@@ -33,6 +34,12 @@ var _ = Describe("Nodes", func() {
 		})
+		It("should able to read NMState resources with cluster-reader user", func() {
+			Eventually(func() error {
+				_, err := cmd.Kubectl("get", "nns", "--as=system:serviceaccount:default:test-user")
```
You don't need to call kubectl, you can use the controller-runtime client instancianting it with Impersonate options https://pkg.go.dev/k8s.io/client-go/rest#Config

oshoval replied:
As we spoke, it would require another client configured as impersonate in order to keep the kube:admin client intact,
and doesn't seem to worth it, we can recalculate later if needed

qinqon commented on test/e2e/handler/nodes_test.go:
```diff
@@ -33,6 +34,12 @@ var _ = Describe("Nodes", func() {
+			Eventually(func() error {
+				_, err := cmd.Kubectl("get", "nns", "--as=system:serviceaccount:default:test-user")
+				return err
+			}, 5*time.Second, 1*time.Second).Should(Succeed())
```
```suggestion
			}, 5*time.Second, time.Second).Should(Succeed())
```

qinqon commented on test/e2e/handler/nodes_test.go:
```diff
@@ -33,6 +34,12 @@ var _ = Describe("Nodes", func() {
 				interfacesNameForNodeEventually(node).Should(ContainElement(primaryNic))
 			}
 		})
+		It("should able to read NMState resources with cluster-reader user", func() {
```
```suggestion
		It("should be able to read NMState resources with cluster-reader role user", func() {
```

qinqon commented on deploy/handler/cluster_role.yaml:
```diff
@@ -0,0 +1,19 @@
+{{if .ClusterReaderExists}}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: nmstate-cluster-reader
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-cluster-reader: "true"
+rules:
+- apiGroups:
+  - nmstate.io
+  resources:
+  - nodenetworkstates
+  - nodenetworkconfigurationpolicies
+  - nodenetworkconfigurationenactments
+  verbs:
+  - get
+  - list
+  - watch
+{{end}}
```
What do you think about removing this from the "role.yaml" and create another role binding to bind this role to the nmstate-handler user ? this way we don't need the conditional since we must deploy it.

oshoval replied:
We need to allow the cluster-reader to read those, not other user, so it won't help
and imo self contained like this
unless I miss anything ?

oshoval commented:
```
+ ./cluster/kubectl.sh rollout status -w -n nmstate ds nmstate-handler
selecting docker as container runtime
Error from server (NotFound): daemonsets.apps "nmstate-handler" not found
+ echo 'Handler haven'\''t turned ready within the given timeout'
Handler haven't turned ready within the given timeout
+ return 1
make: *** [Makefile:175: cluster-sync] Error 1
```
The failures here happen not only on this PR, need to extend the timeout as after few seconds it is fine if you check the cluster manually

oshoval commented:
addressed comments

oshoval commented:
/retest 

rollout timeouts - trying to address at https://github.com/nmstate/kubernetes-nmstate/pull/1069

oshoval commented on test/e2e/operator/nmstate_install_test.go:
```diff
@@ -130,6 +131,40 @@ var _ = Describe("NMState operator", func() {
+	Context("when cluser-reader exists", func() {
+		const (
+			clusterReaderRoleName = "cluster-reader"
+			testUserNamespace     = "default"
```
maybe better on the nmstate NS, so once it is deleted it will clean tests leftovers if any

oshoval commented:
/test pull-kubernetes-nmstate-e2e-handler-k8s

oshoval commented:
/test pull-kubernetes-nmstate-e2e-handler-k8s-future 
/test pull-kubernetes-nmstate-e2e-handler-k8s 

PR to fix the timeout was merged

oshoval commented:
wiill return the Eventually, because it take a moment until there is NNS

creydr commented on test/e2e/operator/nmstate_install_test.go:
```diff
@@ -130,6 +131,41 @@ var _ = Describe("NMState operator", func() {
+		)
+
+		BeforeEach(func() {
+			Expect(createClusterReaderCR(clusterReaderRoleName)).To(Succeed(),
```
This probably fails on testing on Openshift as the cluster role exists already. But you could check in case of an error is it already existed (`errors.IsAlreadyExists(err)`)

creydr commented on test/e2e/operator/nmstate_install_test.go:
```diff
@@ -130,6 +131,41 @@ var _ = Describe("NMState operator", func() {
+
+			Expect(deleteTestUserCRB(testUserBindingName)).To(Succeed())
+			Expect(deleteTestUserSA(testUserNamespace, serviceAccountName)).To(Succeed())
+			Expect(deleteClusterReaderCR(clusterReaderRoleName)).To(Succeed())
```
only delete, if **you** created it

oshoval replied:
all of them ?
We are safe to do so only for the `deleteClusterReaderCR` right ?
since as you said above i will check if its not already exists, and if it exists i won't create nor delete
The two others we know I created because if they are already there, it would fail on the creation.

In case someone interrupts after creation and before deletion, it is better that we will clean those on the run so the next run will be clean.
Their name is unique (has test in it), so it is unlikely someone else created them.

Another point Is that we can remove split those 3 lines to 3 AfterEach so if one fails, it will still perform the others.

wdyt ?

oshoval replied:
> To be more unique, you could also use the `metadata.GenerateName` and let k8s decide on its "final" name.

I prefer deterministic, unless you think its necessary

oshoval commented:
Thanks, addressed comments
Also split to separated `AfterEach` so in case one fails, it will still clean the rest

oshoval commented:
/retest

qinqon commented on test/e2e/operator/nmstate_install_test.go:
```diff
@@ -130,6 +131,54 @@ var _ = Describe("NMState operator", func() {
+			installNMState(defaultOperator.nmstate)
+			eventuallyOperandIsReady(defaultOperator)
+		})
+		AfterEach(func() {
+			uninstallNMStateAndWaitForDeletion(defaultOperator)
+		})
+		AfterEach(func() {
+			Expect(deleteTestUserCRB(testUserBindingName)).To(Succeed())
+		})
+		AfterEach(func() {
+			Expect(deleteTestUserSA(testUserNamespace, serviceAccountName)).To(Succeed())
+		})
+		AfterEach(func() {
+			if clusterReaderCreated {
+				Expect(deleteClusterReaderCR(clusterReaderRoleName)).To(Succeed())
+			}
+		})
```
why Three `AfterEach` ?

oshoval replied:
so if one fails, it will still try to clean the others

qinqon commented:
/lgtm
/approve
